### DISCUSSION TITLE
How do the ssst functions work when given multi-shell data?

### DISCUSSION THREAD

**Discussion by ebrahimebrahim** [dipy/dipy#3037] [Reconstruction]
I have multi-shell dwi data (i.e. multiple b-values besides b=0) and I know that in order to do CSD with this I can use the tools iin `dipy.reconst.mcsd`.

But I am curious what happens if I use tools from `dipy.reconst.csdeconv`, which all seem to be based on single shell data. It seems that I can pass to `dipy.reconst.csdeconv.response_from_mask_ssst` or `dipy.reconst.csdeconv.recursive_response` my DWI data and my gradient table with multiple nonzero b-values, and it computes a response without complaint. The output always seems to be a single response, where I would instead have expected one response for each unique nonzero b-value.

I am struggling to figure out just from the code how this is working.

- Is it actually user error on my part to pass multishell data to these functions? Is the proper usage to actually choose a b value and restrict the data to become single shell before using `dipy.reconst.csdeconv` tools?
- Or is it internally choosing a single shell to work with? From digging through the code I don't see that happening but maybe I have missed it.
- Or am I misunderstanding something and there is in fact some way to use the multi-shell data to generate a single response function to use for CSD??

**Comment by pjsjongsung** [comment]
I can say for sure 2. is not the case. Regardless of whether it is 1 or 3, I believe documentation of the functions should be changed to reduce such misunderstanding/unexpected behavior.

**Comment by arokem** [comment]
The answer is the third option among your answers. There is nothing that
prevents you from determining one response function that would fit the data
in multiple shells and using that response function with multi-shell data.
In this case, we're fitting DTI to the data from multiple shells (
https://github.com/dipy/dipy/blob/24f76c8eb30aa9d256b2e163987139fd0c9bda68/dipy/reconst/csdeconv.py#L1004)
and generating one response function based on that. Does that make sense?

**Reply by ebrahimebrahim** [reply]
Hmm, I am still confused, but this answer helps me phrase my question more specifically.

Fitting _DTI_ to multi-shell data makes sense, but I thought that the actual response function -- the kernel used for spherical deconvolution -- was not a diffusion _tensor_ but rather a diffusion _signal_. Now I see that `dipy.reconst.csdeconv` works with both types of representations, and in each case I am confused for different reasons:

1. When using `dipy.reconst.csdeconv.recursive_response` , the response function returned is in the form of an `AxSymShResponse` object, which I guess is a diffusion _signal_ in spherical harmonic representation. Here I have the confusion from my original question: when given multi-shell data, how is a single diffusion signal being estimated as the response function?  Shouldn't it be one response function for each nonzero b-value? Is one b-value getting selected or what?
2. When using `dipy.reconst.csdeconv.response_from_mask_ssst`, the "response function" returned is in the form of a diffusion _tensor_ (represented in minimal way by two unique eigenvalues and an S0 value, making use of the symmetries of a response function). Fitting multi-shell data to a tensor is fine. The multiple shells would help improve the accuracy/robustness of the fit.. However in order to do spherical deconvolution, we need to have the response function as a diffusion signal. When we feed this response function to `ConstrainedSphericalDeconvModel`, it [generates diffusion signals from the tensor](https://github.com/dipy/dipy/blob/24f76c8eb30aa9d256b2e163987139fd0c9bda68/dipy/reconst/csdeconv.py#L266) to then use in spherical deconvolution. But I am having trouble understanding what it's doing. The generated signal should still depend on b-value. Is it using all the b-values and doing the spherical-deconvolution-least-squares-fit-thing over all the b-values at once to estimate the fiber ODFs?

So does `dipy.reconst.csdeconv.ConstrainedSphericalDeconvModel` effectively do _multi-shell_ CSD when given multi-shell data and a response function generated via `dipy.reconst.csdeconv.response_from_mask_ssst`? And it must be doing some kind of _single-shell_ CSD when given a response function generated via `dipy.reconst.csdeconv.recursive_response`?

**Comment by arokem** [comment]
On Sun, Jan 14, 2024 at 7:00 PM Ebrahim Ebrahim ***@***.***>
wrote:

> Hmm, I am still confused, but this answer helps me phrase my question more
> specifically.
>
> Fitting *DTI* to multi-shell data makes sense, but I thought that the
> actual response function -- the kernel used for spherical deconvolution --
> was not a diffusion *tensor* but rather a diffusion *signal*. Now I see
> that dipy.reconst.csdeconv works with both types of representations, and
> in each case I am confused for different reasons:
>
>    1. When using dipy.reconst.csdeconv.recursive_response , the response
>    function returned is in the form of an AxSymShResponse object, which I
>    guess is a diffusion *signal* in spherical harmonic representation.
>    Here I have the confusion from my original question: when given multi-shell
>    data, how is a single diffusion signal being estimated as the response
>    function? Shouldn't it be one response function for each nonzero b-value?
>    Is one b-value getting selected or what?
>    2. When using dipy.reconst.csdeconv.response_from_mask_ssst, the
>    "response function" returned is in the form of a diffusion *tensor*
>    (represented in minimal way by two unique eigenvalues and an S0 value,
>    making use of the symmetries of a response function). Fitting multi-shell
>    data to a tensor is fine. The multiple shells would help improve the
>    accuracy/robustness of the fit.. However in order to do spherical
>    deconvolution, we need to have the response function as a diffusion signal.
>    When we feed this response function to ConstrainedSphericalDeconvModel,
>    it generates diffusion signals from the tensor
>    <https://github.com/dipy/dipy/blob/24f76c8eb30aa9d256b2e163987139fd0c9bda68/dipy/reconst/csdeconv.py#L266>
>    to then use in spherical deconvolution. But I am having trouble
>    understanding what it's doing. The generated signal should still depend on
>    b-value. Is it using all the b-values and doing the
>    spherical-deconvolution-least-squares-fit-thing over all the b-values at
>    once to estimate the fiber ODFs?
>
> So does dipy.reconst.csdeconv.ConstrainedSphericalDeconvModel effectively
> do *multi-shell* CSD when given multi-shell data and a response function
> generated via dipy.reconst.csdeconv.response_from_mask_ssst? And it must
> be doing some kind of *single-shell* CSD when given a response function
> generated via dipy.reconst.csdeconv.recursive_response?
>

The answer to your number 2 above is "yes". One potential change we could
introduce is to use a more sophisticated model of diffusion for this
response function (for example, a relatively easy tweak would be to use DKI
instead of DTI for multi-shell data).

I am not sure about number 1, but maybe others know better than me.

**Reply by skoudoro** [reply]
Can you answer to the question above @karanphil?

**Reply by ebrahimebrahim** [reply]
Thanks for clarifying about (2)! That's great, it makes sense now. Another suggestion is maybe `response_from_mask_ssst` and `auto_response_ssst` can be renamed since they do handle and work with multishell data. I think ssst stands for "**single shell** single tissue".

Waiting to see if anyone knows about (1) :smiley:

**Reply by arokem** [reply]
Yeah - I agree that's a confusing name. I wonder what others think about that.

**Reply by karanphil** [reply]
As stated in these functions, the DTI fit might be affected by b-values higher than 1200 and the use of such b-values is not recommanded, in particualr for `mask_for_response_ssst`. So typical multi-shell data is not recommanded right away.

Moreover, these functions will indeed accept and handle multi-shell data, but will not produce multi-shell response functions. These names don't stand for what is the input, rather what is the output used for (response function for ssst or msmt-CSD).

**Reply by ebrahimebrahim** [reply]
> As stated in these functions, the DTI fit might be affected by b-values higher than 1200 and the use of such b-values is not recommanded

By the way this is not stated in [the documentation of `mask_for_response_ssst`](https://github.com/dipy/dipy/blob/9153a852a511b07ffc141becfbd9a96ca42e9a90/dipy/reconst/csdeconv.py#L889-L926). In `mask_for_response_msmt` it is stated [in the form of a warning](https://github.com/dipy/dipy/blob/9153a852a511b07ffc141becfbd9a96ca42e9a90/dipy/reconst/mcsd.py#L605).

**Comment by karanphil** [comment]
Unfortunately, I have never used `dipy.reconst.csdeconv.recursive_response`. I can however clarify a few things regarding `dipy.reconst.csdeconv.response_from_mask_ssst`.

It is very important to note that `dipy.reconst.csdeconv.ConstrainedSphericalDeconvModel` **does not** effectively do multi-shell multi-tissue CSD (msmt-CSD) when given multi-shell data and a response function generated via`dipy.reconst.csdeconv.response_from_mask_ssst`. The latter function always returns one single response function describing WM only, and should not be computed from multi-shell data (maybe something to clarify in the documentation).  Thus, it cannot be used to do multi-shell **multi-tissue** CSD. The keyword here is multi-tissue. What can be done however, is computing a single response function from a selected shell of your multi-shell data (preferably a shell between 500 and 1500 s/mm² for optimal DTI fit), and then performing multi-shell **single-tissue** CSD (what I call msst-CSD). This is used for instance in the Tractoflow pipeline (https://tractoflow-documentation.readthedocs.io/en/latest/).

There is an important distinction to make between multi-shell and multi-tissue, as the former does not implie the latter. The "usual" multi-shell CSD requires a response function per tissue per unique b-value and includes at least two tissues. This cannot be done with `dipy.reconst.csdeconv.response_from_mask_ssst`, rather with `dipy.reconst.mcsd.response_from_mask_msmt`.

I hope it is somewhat clear and helpful!

**Reply by karanphil** [reply]
In order to use your multi-shell data with `dipy.reconst.csdeconv`, I would go for the msst-CSD route. This means I would do the following steps:

1. Compute the mask for response function ssst (WM) using only the "DTI" shells, so b-values between 500 and 1500 s/mm², with `mask_for_response_ssst`.
2. Compute the ssst response function from the mask, using again only the "DTI" shells, with `dipy.reconst.csdeconv.response_from_mask_ssst`.
3. Compute the msst-CSD using all your data as input, but the ssst response function, with `dipy.reconst.csdeconv`.

**Reply by ebrahimebrahim** [reply]
I see, yes your answer makes sense.

>  Thus, it cannot be used to do multi-shell multi-tissue CSD.

So is it user error to do so anyway? Because you totally can feed multi-shell data into `mask_for_response_ssst` and `response_from_mask_ssst` and there is no complaint. If it's always an error to use it in this way then maybe it should raise an error and suggest the user to cut the input down to one b value... am I understanding correctly?

> In order to use your multi-shell data with dipy.reconst.csdeconv, I would go for the msst-CSD route. This means I would do the following steps:

In general given multi-shell data would you prefer that ms**s**t-csd strategy that you listed, or would you prefer using the tools in `dipy.reconst.mcsd` to do ms**m**t-csd?

**Reply by karanphil** [reply]
Yes, a check on the number of shells could be useful in that case, at least to raise a warning. I think multiple shells below 1500 s/mm^2 could theoretically still be used to increase the precision of the DTI fit when few directions are available, but I am not sure.

It might depend on the end goal, but I am a believer of msmt-CSD in general. The msst-CSD strategy is reliable and most people in my lab use it to produce fODFs used for tractography. The major difference is access to the signal fractions of the various tissues like WM, GM and CSF, and perhaps better reconstruction in partial volume (with msmt-CSD). This is my experience, there is not much literature about this.

**Reply by ebrahimebrahim** [reply]
Thanks for sharing your perspective :smiley:
